File: Stage20/C/Problem2580.c
#include <stdio.h>

int board[9][9];
int blanks[81][2];
int blank_count = 0;
int solved = 0;

void dfs(int idx) {
    if (idx == blank_count) {
        solved = 1;

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++)
                printf("%d ", board[i][j]);
            printf("\n");
        }

        return;
    }

    int r = blanks[idx][0];
    int c = blanks[idx][1];

    for (int num = 1; num <= 9 && !solved; num++) {
        if (check(r, c, num)) {
            board[r][c] = num;
            dfs(idx + 1);
            board[r][c] = 0;
        }
    }
}

int check(int r, int c, int num) {
    for (int i = 0; i < 9; i++)
        if (board[r][i] == num || board[i][c] == num)
            return 0;

    int sr = (r / 3) * 3;
    int sc = (c / 3) * 3;
    for (int i = sr; i < sr + 3; i++)
        for (int j = sc; j < sc + 3; j++)
            if (board[i][j] == num)
                return 0;

    return 1;
}

int main(void) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            scanf("%d", &board[i][j]);
            if (board[i][j] == 0) {
                blanks[blank_count][0] = i;
                blanks[blank_count][1] = j;
                blank_count++;
            }
        }
    }

    dfs(0);
    return 0;
}

File: Stage20/Python/Problem2580.py
import sys
input = sys.stdin.readline

board = []
blanks = []
row = [[False] * 10 for _ in range(9)]
col = [[False] * 10 for _ in range(9)]
box = [[False] * 10 for _ in range(9)]

def init():
    global board, blanks
    board = [list(map(int, input().split())) for _ in range(9)]

    for i in range(9):
        for j in range(9):
            num = board[i][j]
            if num == 0:
                blanks.append((i, j))
            else:
                row[i][num] = True
                col[j][num] = True
                box[(i // 3) * 3 + j // 3][num] = True

def dfs(idx):
    if idx == len(blanks):
        for r in board:
            print(*r)
        exit(0)

    i, j = blanks[idx]
    b = (i // 3) * 3 + j // 3

    for num in range(1, 10):
        if not row[i][num] and not col[j][num] and not box[b][num]:
            board[i][j] = num
            row[i][num] = col[j][num] = box[b][num] = True

            dfs(idx + 1)

            board[i][j] = 0
            row[i][num] = col[j][num] = box[b][num] = False

def main():
    init()
    dfs(0)

if __name__ == "__main__":
    main()